nilearn/decomposition/dict_learning.py

```diff
@@ -0,0 +1,201 @@
+    def __init__(self, mask=None, n_components=20,
+                 smoothing_fwhm=6,
+                 standardize=True,
+                 random_state=0,
+                 target_affine=None, target_shape=None,
+                 low_pass=None, high_pass=None, t_r=None,
+                 alpha=1,
+                 n_iter=1000,
+                 # Common options
+                 memory=Memory(cachedir=None), memory_level=0,
+                 n_jobs=1, verbose=0,
+                 dict_init=None
+                 ):
+        SinglePCA.__init__(self,
+                           mask=mask, memory=memory, memory_level=memory_level,
+                           n_jobs=n_jobs, verbose=verbose,
+                           n_components=n_components, smoothing_fwhm=smoothing_fwhm,
+                           target_affine=target_affine, target_shape=target_shape,
+                           random_state=random_state, high_pass=high_pass, low_pass=low_pass,
+                           t_r=t_r,
+                           standardize=standardize,
+                           do_cca=False)
+        self.n_iter = n_iter
+        self.alpha = alpha
+        self.dict_init = dict_init
+
+    def _init_dict(self, imgs, y=None, confounds=None):
+        if self.dict_init is not None:
+            self.dict_init_ = self.dict_init
+        else:
+            # Perform a CanICA initialization using SinglePCA parameters
+            parameters = get_params(SinglePCA, self)
+            # Efficient CanICA requires CCA
+            parameters['do_cca'] = True
+            canica = CanICA(threshold=float(self.n_components), n_init=1,
+                            **parameters)
+            canica.fit(imgs, y=y, confounds=confounds)
+
+            ridge = Ridge(alpha=1e-6, fit_intercept=None)
+            ridge.fit(canica.components_.T, self.data_flat_.T)
+            self.dict_init_ = ridge.coef_.T
+            S = np.sqrt(np.sum(self.dict_init_ ** 2, axis=0))
+            self.dict_init_ /= S[np.newaxis, :]
+
+    def fit(self, imgs, y=None, confounds=None):
+        """Compute the mask and the ICA maps across subjects
+
+        Parameters
+        ----------
+        imgs: list of Niimg-like objects
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            Data on which PCA must be calculated. If this is a list,
+            the affine is considered the same for all.
+
+        confounds: CSV file path or 2D matrix
+            This parameter is passed to nilearn.signal.clean. Please see the
+            related documentation for details
+        """
+        if self.n_iter == 'auto':
+            # ceil(self.data_fat.shape[0] / self.batch_size)
+            self.n_iter = (self.data_flat_.shape[0] - 1) / self.batch_size + 1
+
+        if self.verbose:
+            print('[DictLearning] Loading data')
+        SinglePCA.fit(self, imgs, y=y, confounds=confounds)
+        self.data_flat_ = np.concatenate(self.components_)
```

Comment:
If you look at the original code, there is an efficient concatenation after the computation of session_pca. Maybe you should systematically concatenate afterward.